rs-matter/src/utils/ifmutex.rs

```diff
@@ -0,0 +1,217 @@
+ *    limitations under the License.
+ */
+
+//! A variation of the `embassy-sync` async mutex that only locks the mutex if a certain condition on the content of the data holds true.
+//! Check `embassy_sync::Mutex` for the original unconditional implementation.
+use core::cell::UnsafeCell;
+use core::ops::{Deref, DerefMut};
+
+use embassy_sync::blocking_mutex::raw::RawMutex;
+
+use super::signal::Signal;
+
+/// Error returned by [`Mutex::try_lock`]
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct TryLockError;
+
+/// Async mutex with conditional locking based on the data inside the mutex.
+/// Check `embassy_sync::Mutex` for the original unconditional implementation.
+pub struct IfMutex<M, T>
+where
+    M: RawMutex,
+    T: ?Sized,
+{
+    state: Signal<M, bool>,
+    inner: UnsafeCell<T>,
+}
+
+unsafe impl<M: RawMutex + Send, T: ?Sized + Send> Send for IfMutex<M, T> {}
+unsafe impl<M: RawMutex + Sync, T: ?Sized + Send> Sync for IfMutex<M, T> {}
+
+/// Async mutex.
+impl<M, T> IfMutex<M, T>
+where
+    M: RawMutex,
+{
+    /// Create a new mutex with the given value.
+    #[inline(always)]
+    pub const fn new(value: T) -> Self {
+        Self {
+            state: Signal::<M, _>::new(false),
+            inner: UnsafeCell::new(value),
+        }
+    }
+}
+
+impl<M, T> IfMutex<M, T>
+where
+    M: RawMutex,
+    T: ?Sized,
+{
+    /// Lock the mutex.
+    ///
+    /// This will wait for the mutex to be unlocked if it's already locked.
+    pub async fn lock(&self) -> IfMutexGuard<'_, M, T> {
+        self.lock_if(|_| true).await
+    }
+
+    /// Lock the mutex.
+    ///
+    /// This will wait for the mutex to be unlocked if it's already locked _and_ for the provided condition on the data to become true.
+    pub async fn lock_if<F>(&self, f: F) -> IfMutexGuard<'_, M, T>
+    where
+        F: Fn(&T) -> bool,
+    {
+        self.state
+            .wait(|locked| {
+                if !*locked && f(unsafe { &*self.inner.get() }) {
+                    *locked = true;
+
+                    Some(())
+                } else {
+                    None
+                }
+            })
+            .await;
+
+        IfMutexGuard { mutex: self }
+    }
+
+    /// Waits for the mutex to become unlocked and then executes the provided closure.
+    /// Will become ready only when the callback closure returns a `Some` result.
+    pub async fn with<F, R>(&self, mut f: F) -> R
+    where
+        F: FnMut(&mut T) -> Option<R>,
+    {
+        let result = self
+            .state
+            .wait(|locked| {
+                if !*locked {
+                    if let Some(result) = f(unsafe { &mut *self.inner.get() }) {
+                        *locked = true;
+                        return Some(result);
+                    }
+                }
+
+                None
+            })
+            .await;
+
+        let _ = IfMutexGuard { mutex: self };
```

Comment:
a comment here that we use this for the drop implementation to unlock the locked state above may be useful.